chain/index/events.go

```diff
@@ -236,48 +238,98 @@ func loadExecutedMessages(ctx context.Context, cs ChainStore, recomputeTipSetSta
-	case f.MinHeight >= 0 && f.MinHeight == f.MaxHeight:
-		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
-		if err != nil {
-			if err == ErrNotFound {
-				// this means that this is a null round and there exist no events for this epoch
-				return nil
-			}
+// checkRangeIndexedStatus verifies if a range of heights is indexed.
+// It checks for the existence of non-null rounds at the range boundaries.
+func (si *SqliteIndexer) checkRangeIndexedStatus(ctx context.Context, f *EventFilter) error {
+	minHeight := f.MinHeight
+	maxHeight := f.MaxHeight
 
-			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+	// Find the first non-null round in the range
+	startCid, err := si.findFirstNonNullRound(ctx, &minHeight, maxHeight)
+	if err != nil {
+		return xerrors.Errorf("failed to find first non-null round: %w", err)
+	}
+
+	// If all rounds are null, consider the range valid
+	if startCid == nil {
+		return nil
+	}
+
+	// Find the last non-null round in the range
+	endCid, err := si.findLastNonNullRound(ctx, &maxHeight, minHeight)
+	if err != nil {
+		if errors.Is(err, ErrRangeInFuture) {
+			return xerrors.Errorf("range end is in the future: %w", err)
 		}
-	default:
-		// This function distinguishes between two scenarios:
-		// 1. Missing events: The requested tipset is not present in the Index (an error condition).
-		// 2. Valid case: The tipset exists but contains no events (a normal situation).
-		// Currently, this distinction is only made for the common use case where a user requests events for a single tipset.
-		// TODO: Implement this functionality for a range of tipsets. This is expensive and not a common use case so it's deferred for now.
+		return xerrors.Errorf("failed to find last non-null round: %w", err)
+	}
+
+	// If all rounds are null, consider the range valid
+	if endCid == nil {
 		return nil
 	}
 
-	// If we couldn't determine a specific tipset, return ErrNotFound
-	if tipsetKeyCid == nil {
-		return ErrNotFound
+	// Check indexing for start and end tipsets
+	if err := si.checkTipsetByKeyCid(ctx, startCid, minHeight); err != nil {
+		return err
 	}
 
-	// Check if the determined tipset is indexed
-	if exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid); err != nil {
-		return xerrors.Errorf("failed to check if tipset is indexed: %w", err)
-	} else if exists {
-		return nil // Tipset is indexed
+	if err := si.checkTipsetByKeyCid(ctx, endCid, maxHeight); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// checkTipsetByKeyCid checks if a tipset identified by its key CID is indexed.
+func (si *SqliteIndexer) checkTipsetByKeyCid(ctx context.Context, tipsetKeyCid []byte, height abi.ChainEpoch) error {
+	exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid)
+	if err != nil {
+		return xerrors.Errorf("failed to check if tipset at height %d is indexed: %w", height, err)
+	}
+
+	if exists {
+		return nil // null round
+	}
+
+	return ErrNotFound // tipset is not indexed
+}
```

Comment:
I'm not really sure the `height` is necessary here, but the main thing I want to do is clarify what this is for. How about we rename this one `checkTipsetIndexedStatus` since you've dropped it form its parent


```suggestion
func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, tipsetKeyCid []byte, height abi.ChainEpoch) error {
	exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid)
	if err != nil {
		return xerrors.Errorf("failed to check if tipset at height %d is indexed: %w", height, err)
	} else if exists {
		return nil // has been indexed
	}

	return ErrNotFound
}
```